### PR TITLE
feat(backend): add deployment logs to the backend

### PR DESCRIPTION
**Pull request by hanshal101**
# Description 

In this PR I have added the backend code for getting the deployment logs.

## Method
Here I have researched on how to fetch the metrics/logs from the kubernetes API/ client-go. But there was no way to fetch this, but we can fetch logs of a particular pod. Hence here I have fetched the logs of all the available pods available/present in that particular deployment.

### CODE REVIEW

**Review by petar-cvit — CHANGES_REQUESTED** [review]

**Comment by petar-cvit** [comment]
Does the current solution just concatenate the logs from multiple pods? We need to sort them by time so they don't confuse the person reading those

**Comment by hanshal101** [comment]
I'll need to check this as my current modules don't have any logs to show, Do you have any idea how can I make them to show some logs, either messing with them will do😅

**Comment by petar-cvit** [comment]
I have an image I use for testing that prints a log `hello Polyfem` every couple of seconds. You can use `petarcvit223/logs:latest`

**Comment by hanshal101** [comment]
So this is the response from the current deployment logs:
```
[
    "Try setting the name with NAME env var\nHello polyfem 0\nHello polyfem 1\nHello polyfem 2\nHello polyfem 3\nHello polyfem 4\nHello polyfem 5\nHello polyfem 6\nHello polyfem 7\nHello polyfem 8\nHello polyfem 9\nHello polyfem 10\nHello polyfem 11\nHello polyfem 12\nHello polyfem 13\nHello polyfem 14\nHello polyfem 15\nHello polyfem 16\nHello polyfem 17\nHello polyfem 18\nHello polyfem 19\nHello polyfem 20\nHello polyfem 21\nHello polyfem 22\nHello polyfem 23\nHello polyfem 24\nHello polyfem 25\nHello polyfem 26\nHello polyfem 27\nHello polyfem 28\nHello polyfem 29\nHello polyfem 30\nHello polyfem 31\nHello polyfem 32\nHello polyfem 33\nHello polyfem 34\nHello polyfem 35\nHello polyfem 36\n",
    "Try setting the name with NAME env var\nHello polyfem 0\nHello polyfem 1\nHello polyfem 2\nHello polyfem 3\nHello polyfem 4\nHello polyfem 5\nHello polyfem 6\nHello polyfem 7\nHello polyfem 8\nHello polyfem 9\nHello polyfem 10\nHello polyfem 11\nHello polyfem 12\nHello polyfem 13\nHello polyfem 14\nHello polyfem 15\nHello polyfem 16\nHello polyfem 17\nHello polyfem 18\nHello polyfem 19\nHello polyfem 20\nHello polyfem 21\nHello polyfem 22\nHello polyfem 23\nHello polyfem 24\nHello polyfem 25\nHello polyfem 26\nHello polyfem 27\nHello polyfem 28\nHello polyfem 29\nHello polyfem 30\nHello polyfem 31\nHello polyfem 32\nHello polyfem 33\nHello polyfem 34\nHello polyfem 35\nHello polyfem 36\nHello polyfem 37\nHello polyfem 38\nHello polyfem 39\nHello polyfem 40\nHello polyfem 41\nHello polyfem 42\nHello polyfem 43\nHello polyfem 44\nHello polyfem 45\nHello polyfem 46\nHello polyfem 47\nHello polyfem 48\nHello polyfem 49\nHello polyfem 50\nHello polyfem 51\nHello polyfem 52\nHello polyfem 53\nHello polyfem 54\nHello polyfem 55\nHello polyfem 56\nHello polyfem 57\n"
]
```

**Comment by petar-cvit** [comment]
Just tried it, and it seems like we have a bug in `GetPodLogs` which returns an array of a single string that has logs concatenated with `\n`. Can you fix that method as well?

**Comment by hanshal101** [comment]
Yup I can do this, I think I can solve this problem by `bufio`

**Comment by hanshal101** [comment]
image
I think this looks good now, should I change the same for `Deployments`

**Comment by petar-cvit** [comment]
Yeah, looks good! Now you just need to aggregate those for Deployments

**Comment by hanshal101** [comment]
Well i just tried the existing code for `deployment logs` its the same, the out put looks like this:
```
[
    "Hello polyfem 424",
    "Hello polyfem 425",
    "Hello polyfem 426",
    "Hello polyfem 427",
    "Hello polyfem 428",
    "Hello polyfem 429",
    "Hello polyfem 430",
    "Hello polyfem 431",
    "Hello polyfem 432",
    "Hello polyfem 433",
    "Hello polyfem 434",
    "Hello polyfem 435",
    "Hello polyfem 436",
    "Hello polyfem 437",
    "Hello polyfem 438",
    "Hello polyfem 439",
    "Hello polyfem 440",
    "Hello polyfem 441",
    "Hello polyfem 442",
    "Hello polyfem 443",
    "Hello polyfem 444",
    "Hello polyfem 445",
    "Hello polyfem 446",
    "Hello polyfem 447",
    "Hello polyfem 448",
    "Hello polyfem 449",
    "Hello polyfem 450",
    "Hello polyfem 451",
    "Hello polyfem 452",
    "Hello polyfem 453",
    "Hello polyfem 454",
    "Hello polyfem 455",
    "Hello polyfem 456",
    "Hello polyfem 457",
    "Hello polyfem 458",
    "Hello polyfem 459",
    "Hello polyfem 460",
    "Hello polyfem 461",
    "Hello polyfem 462",
    "Hello polyfem 463",
    "Hello polyfem 464",
    "Hello polyfem 465",
    "Hello polyfem 466",
    "Hello polyfem 467",
    "Hello polyfem 468",
    "Hello polyfem 469",
    "Hello polyfem 470",
    "Hello polyfem 471",
    "Hello polyfem 472",
    "Hello polyfem 473",
    "Hello polyfem 474",
    "Hello polyfem 475",
    "Hello polyfem 476",
    "Hello polyfem 477",
    "Hello polyfem 478",
    "Hello polyfem 479",
    "Hello polyfem 480",
    "Hello polyfem 481",
    "Hello polyfem 482",
    "Hello polyfem 483",
    "Hello polyfem 484",
    "Hello polyfem 485",
    "Hello polyfem 486",
    "Hello polyfem 487",
    "Hello polyfem 488",
    "Hello polyfem 489",
    "Hello polyfem 490",
    "Hello polyfem 491",
    "Hello polyfem 492",
    "Hello polyfem 493",
    "Hello polyfem 494",
    "Hello polyfem 495",
    "Hello polyfem 496",
    "Hello polyfem 497",
    "Hello polyfem 498",
    "Hello polyfem 499",
    "Hello polyfem 500",
    "Hello polyfem 501",
    "Hello polyfem 502",
    "Hello polyfem 503",
    "Hello polyfem 504",
    "Hello polyfem 505",
    "Hello polyfem 506",
    "Hello polyfem 507",
    "Hello polyfem 508",
    "Hello polyfem 509",
    "Hello polyfem 510",
    "Hello polyfem 511",
    "Hello polyfem 512",
    "Hello polyfem 513",
    "Hello polyfem 514",
    "Hello polyfem 515",
    "Hello polyfem 516",
    "Hello polyfem 517",
    "Hello polyfem 518",
    "Hello polyfem 519",
    "Hello polyfem 520",
    "Hello polyfem 521",
    "Hello polyfem 522",
    "Hello polyfem 523",
    "Hello polyfem 423",
    "Hello polyfem 424",
    "Hello polyfem 425",
    "Hello polyfem 426",
    "Hello polyfem 427",
    "Hello polyfem 428",
    "Hello polyfem 429",
    "Hello polyfem 430",
    "Hello polyfem 431",
    "Hello polyfem 432",
    "Hello polyfem 433",
    "Hello polyfem 434",
    "Hello polyfem 435",
    "Hello polyfem 436",
    "Hello polyfem 437",
    "Hello polyfem 438",
    "Hello polyfem 439",
    "Hello polyfem 440",
    "Hello polyfem 441",
    "Hello polyfem 442",
    "Hello polyfem 443",
    "Hello polyfem 444",
    "Hello polyfem 445",
    "Hello polyfem 446",
    "Hello polyfem 447",
    "Hello polyfem 448",
    "Hello polyfem 449",
    "Hello polyfem 450",
    "Hello polyfem 451",
    "Hello polyfem 452",
    "Hello polyfem 453",
    "Hello polyfem 454",
    "Hello polyfem 455",
    "Hello polyfem 456",
    "Hello polyfem 457",
    "Hello polyfem 458",
    "Hello polyfem 459",
    "Hello polyfem 460",
    "Hello polyfem 461",
    "Hello polyfem 462",
    "Hello polyfem 463",
    "Hello polyfem 464",
    "Hello polyfem 465",
    "Hello polyfem 466",
    "Hello polyfem 467",
    "Hello polyfem 468",
    "Hello polyfem 469",
    "Hello polyfem 470",
    "Hello polyfem 471",
    "Hello polyfem 472",
    "Hello polyfem 473",
    "Hello polyfem 474",
    "Hello polyfem 475",
    "Hello polyfem 476",
    "Hello polyfem 477",
    "Hello polyfem 478",
    "Hello polyfem 479",
    "Hello polyfem 480",
    "Hello polyfem 481",
    "Hello polyfem 482",
    "Hello polyfem 483",
    "Hello polyfem 484",
    "Hello polyfem 485",
    "Hello polyfem 486",
    "Hello polyfem 487",
    "Hello polyfem 488",
    "Hello polyfem 489",
    "Hello polyfem 490",
    "Hello polyfem 491",
    "Hello polyfem 492",
    "Hello polyfem 493",
    "Hello polyfem 494",
    "Hello polyfem 495",
    "Hello polyfem 496",
    "Hello polyfem 497",
    "Hello polyfem 498",
    "Hello polyfem 499",
    "Hello polyfem 500",
    "Hello polyfem 501",
    "Hello polyfem 502",
    "Hello polyfem 503",
    "Hello polyfem 504",
    "Hello polyfem 505",
    "Hello polyfem 506",
    "Hello polyfem 507",
    "Hello polyfem 508",
    "Hello polyfem 509",
    "Hello polyfem 510",
    "Hello polyfem 511",
    "Hello polyfem 512",
    "Hello polyfem 513",
    "Hello polyfem 514",
    "Hello polyfem 515",
    "Hello polyfem 516",
    "Hello polyfem 517",
    "Hello polyfem 518",
    "Hello polyfem 519",
    "Hello polyfem 520",
    "Hello polyfem 521",
    "Hello polyfem 522"
]
```
I think this looks good, what do you say?

**Comment by petar-cvit** [comment]
Yeah, this would solve the issue we had with `GetPodLogs`. Does it still look good on the UI?

**Comment by hanshal101** [comment]
I am not really sure of the UI, I tried setting my UI on my machine and there is some issue regarding `npm`.
Can you try on yours?

**Comment by petar-cvit** [comment]
You can try `yarn install` and then `yarn start`

Sure, I can try on mine as well. Can you push your changes?

**Comment by hanshal101** [comment]
What to add more over here?

File: cyclops-ctrl/internal/cluster/k8sclient/client.go [hunk 1]
```diff
@@ -191,6 +193,33 @@ func (k *KubernetesClient) GetPodLogs(namespace, container, name string, numLogs
...
 
+func (k *KubernetesClient) GetDeploymentLogs(namespace, deployment string, numLogs *int64) ([]string, error) {
+	deploymentClient := k.clientset.AppsV1().Deployments(namespace)
+	deploymentObj, err := deploymentClient.Get(context.Background(), deployment, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var logs []string
+	for key, value := range deploymentObj.Spec.Selector.MatchLabels {
```
[the remark this answers is not in the document]
Comment (petar-cvit) [reply cyclops-ctrl/internal/cluster/k8sclient/client.go:198]: Can you address this as well? You can get podList for a deployment like this:
```go
pods, err := k.clientset.CoreV1().Pods(deployment.Namespace).List(context.Background(), metav1.ListOptions{
	LabelSelector: labels.Set(deployment.Spec.Selector.MatchLabels).String(),
})
if err != nil {
	return nil, err
}
```

**Comment by petar-cvit** [comment]
Can you address [this](https://github.com/cyclops-ui/cyclops/pull/232#discussion_r1585257238)?
Also, you will need to sort the logs by time as described [here](https://github.com/cyclops-ui/cyclops/pull/232#issuecomment-2082922567)

**Comment by petar-cvit** [comment]
solves #157

**Comment by petar-cvit** [comment]
Looks good! Can you add the same thing for Statefulsets?

**Comment by hanshal101** [comment]
I have done the logs for StatefulSets too but couldn't verify it. Can you guide me on how to verify this.

**Comment by petar-cvit** [comment]
Hey @hanshal101, I created a test template on a branch. You can try it with:
```
repo: https://github.com/cyclops-ui/templates
path: demo
version: test-statefulset-logs
```
Screenshot 2024-05-04 at 23 08 13
You can deploy a statefulset using it and test the endpoint

**Comment by hanshal101** [comment]
Looks good to me:
image

**Comment by hanshal101** [comment]
Rebased and Refactored !

**Review by petar-cvit — APPROVED** [review]
LGTM

**Pull request merged**